### PR TITLE
login: Allow bypass with "soft" requirements

### PR DESCRIPTION
**Pull request by garrett**
All CSS is now considered "soft", meaning Cockpit will not look right.

Broken CSS may cause components to not be visible or legible.

JavaScript is still mandatory, as functionality breaks without.

Partially addresses #17756

### CODE REVIEW

**Comment by garrett** [comment]
I moved the block around to make it work with the username/password + button. Showing both at the same time looked odd, so I adjusted the style so it would all work together. Additionally, I changed the button to be a "warning" state and changed the text to "Log in anyway". (Better string suggestions are welcome.)

Notably: This does _not_ have a disclaimer that things will be broken.

**Comment by garrett** [comment]
Desktop, dark mode:

image

Desktop, light mode:

image

Mobile, dark:

image


Mobile, light:

image

**Comment by garrett** [comment]
The cancel button here is odd. Perhaps I should add that to the hide part when bypass is active.

**Comment by garrett** [comment]
Force pushed with these changes:

1. Also hide the cancel button when bypassing the soft error
2. Only show separator when there's an error

File: pkg/static/login.js [hunk 1]
```diff
@@ -164,17 +164,26 @@
...
-            document.body.classList.add("brand-unsupported-browser");
+            document.body.classList.add("unsupported-browser");
+
+            if (bypass) {
+                show("#login", "#login-details");
+                hide("#get-out-link");
+                id("login-button").classList.remove("pf-m-primary");
+                id("login-button").classList.add("pf-m-warning");
+                id("login-button-text").textContent = _("Log in anyway");
+            } else {
```
Comment (cockpituous) [inline pkg/static/login.js:184]: These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20220927-095522-73b8faa5-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

Comment (allisonkarlitskaya) [reply pkg/static/login.js:184]: Gonna be interesting trying to figure out how to test that :)

**Review by allisonkarlitskaya — COMMENTED** [review]
I don't have a strong opinion on this in terms of if we ought to do it or not, but one comment: I was imagining something more like the "bad certificate" page approach where we show what looks like a hard error message but have some little button underneath, possibly hidden behind an expander, saying "continue anyway".

I don't expect that we're going to have gotten this wrong in the usual case, and I don't expect that the user is going to want to login, given the issue.  I just want to give an escape hatch.

File: pkg/static/login.js (hunk 1 above)
Comment (garrett) [reply pkg/static/login.js:184]: Yeah. I don't know how. I guess I could split that chunk out to a function... but it's kind of the most actionable part of a function already. I suppose testing the function it's in with a "bypass" value could work... with pixel tests? But I have no clue how do to most of this.

File: pkg/static/login.js [hunk 2]
```diff
@@ -205,7 +214,7 @@
             const args = [].join.call(arguments, ": ");
 
             if (!window.CSS || !window.CSS.supports.apply(this, arguments)) {
-                disableLogin(args);
+                disableLogin(args, "bypass");
```
Comment (marusak) [inline pkg/static/login.js:217]: This is odd pattern to pass string "bypass" into "bypass" argument. Why not just `true`?

File: pkg/static/login.js (hunk 1 above)
Comment (marusak) [reply pkg/static/login.js:184]: can we override browser defined objects? like `window.replaceAll = undefined` and then this would trigger our check?

File: pkg/static/login.js (hunk 2 above)
Comment (marusak) [reply pkg/static/login.js:217]: and we want this only for css not also for js?

Comment (garrett) [reply pkg/static/login.js:217]: Using booleans in function API is usually considered a "code smell" as it's not obvious what's going on.

It's written about in programming books and all over the Internet, but here's a pretty decent blog post on the subject:

https://alexkondov.com/should-you-pass-boolean-to-functions/

Comment (marusak) [reply pkg/static/login.js:217]: > JavaScript is still mandatory, as functionality breaks without.

found the answer in your original comment

Comment (garrett) [reply pkg/static/login.js:217]: Yeah, there's a possibility of optional JS, but we'd have to write defensively, and I'm pretty sure we don't take those things into account. There's no need to support browsers that are a decade old. And, even if we wanted to, they're wildly insecure and our users shouldn't use such an old browser.

We're slower on the uptake of new JavaScript than new CSS (lately).

Things will definitely be broken without proper CSS support... and the we don't generally adopt something new for a year or two anyway (although PF sometimes does sooner, like in the :is()/:where() case).

So if we're going to add a bypass, then, yes, we'd only want that on the CSS and not the JS, unless everyone changed the way they right their JSX and/or we adjust babel specifically for JS... so I figured it's easier and safer to just allow a bypass if only a CSS check fails. (CSS checks come after JS checks for this reason also.)

In other words: There's a big difference between things being completely broken and stuff not looking right (even sometimes horribly wrong or just missing). Letting people through when things are just off-looking or with messed up layouts is OK, as long as we do warn them that their browser doesn't pass the check. But having a browser fail the check for JS and letting them in to be greeted by a non-functional Cockpit is something else entirely.

Therefore, if we just say "if CSS fails, then you can get to it if you really want to" is probably the most correct approach.

**Comment by garrett** [comment]
I'm currently working on @allisonkarlitskaya's suggestion of the expander section. I have some ideas that are now mostly implemented. I plan on pushing up what I have soon. (Just letting you all know there's more incoming.)

The actual code that does stuff won't change too much.

**Comment by garrett** [comment]
I'm in the process of converting this to an expander. It's mostly there, but some things are broken. I'll continue tomorrow. Converted to draft to signify that this isn't done. Also, the commit has WIP on it for now.

File: pkg/static/login.js [hunk 3]
```diff
@@ -164,17 +164,31 @@
...
 
             if (window.console)
                 console.warn(errorString);
             id("login-error-message").textContent = errorString;
-            hide("#login", "#login-details");
             show("#unsupported-browser", "#error-group");
-            document.body.classList.add("brand-unsupported-browser");
+            document.body.classList.add("unsupported-browser");
+
+            if (bypass) {
```
Comment (cockpituous) [inline pkg/static/login.js:178]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20220927-160506-c3e00978-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 4]
```diff
@@ -164,17 +164,31 @@
...
+                show("#login", "#login-details");
+                hide("#get-out-link");
+                // Change the state of the button from primary to warning
+                id("login-button").classList.remove("pf-m-primary");
+                id("login-button").classList.add("pf-m-warning");
+                // Change the text of the log in button
+                id("login-button-text").textContent = _("Log in anyway");
+                // Reparent login form to the expander
+                id("login-override-content").appendChild(id("login"));
+            } else {
```
Comment (cockpituous) [inline pkg/static/login.js:189]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20220927-160506-c3e00978-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 5]
```diff
@@ -164,17 +164,42 @@
     }
 
     function requisites() {
-        function disableLogin(name) {
+        function showBypass(bypass) {
+            if (bypass) {
+                // Selectively show and hide elements
+                show("#login", "#login-details", "#login-override");
+                hide("#get-out-link");
```
Comment (cockpituous) [inline pkg/static/login.js:171]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20220929-151742-405dc998-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 6]
```diff
@@ -164,17 +164,42 @@
...
+
+                // Reparent login form to the expander
+                id("login-override-content").appendChild(id("login"));
```
Comment (cockpituous) [inline pkg/static/login.js:174]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20220929-151742-405dc998-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 7]
```diff
@@ -164,17 +164,42 @@
...
+
+                // Change the state of the button from primary to warning
+                id("login-button").classList.add("pf-m-warning");
```
Comment (cockpituous) [inline pkg/static/login.js:177]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20220929-151742-405dc998-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 8]
```diff
@@ -164,17 +164,42 @@
...
+
+                // Render a "helper text" warning above the log in button
+                document.querySelector("#login .login-actions").insertAdjacentHTML(
+                    "beforebegin",
+                    "<div class='pf-c-helper-text pf-m-warning' id='bypass-warning'>" +
+                    _("Cockpit might not render correctly in your browser") +
+                    "</div>"
```
Comment (cockpituous) [inline pkg/static/login.js:184]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20220929-151742-405dc998-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 9]
```diff
@@ -164,17 +164,42 @@
...
+                );
+            } else {
```
Comment (cockpituous) [inline pkg/static/login.js:186]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20220929-151742-405dc998-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by garrett** [comment]
I don't know how we'd test this and would need assistance in that area. Can someone write tests for this please?

**Comment by garrett** [comment]
I still need assistance with the tests for this PR.

Meanwhile: rebased.

File: pkg/static/login.js (hunk 5 above)
Comment (cockpituous) [inline pkg/static/login.js:171]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221017-161558-ae4aa609-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js (hunk 6 above)
Comment (cockpituous) [inline pkg/static/login.js:174]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221017-161558-ae4aa609-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js (hunk 7 above)
Comment (cockpituous) [inline pkg/static/login.js:177]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221017-161558-ae4aa609-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js (hunk 8 above)
Comment (cockpituous) [inline pkg/static/login.js:184]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221017-161558-ae4aa609-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js (hunk 9 above)
Comment (cockpituous) [inline pkg/static/login.js:186]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221017-161558-ae4aa609-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 10]
```diff
@@ -205,13 +230,13 @@
             const args = [].join.call(arguments, ": ");
 
             if (!window.CSS || !window.CSS.supports.apply(this, arguments)) {
-                disableLogin(args);
+                disableLogin(args, "bypass");
```
Comment (cockpituous) [inline pkg/static/login.js:233]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221017-161558-ae4aa609-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 11]
```diff
@@ -221,11 +246,17 @@
...
                req("replaceAll", String.prototype) &&
                req("finally", Promise.prototype) &&
-               req("supports", window.CSS) &&
-               css("display", "flex") &&
-               css("display", "grid") &&
-               css("selector(test)") &&
-               css("selector(:is(*):where(*))");
+               req("supports", window.CSS);
+
+        if (hard_req) {
```
Comment (cockpituous) [inline pkg/static/login.js:251]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221017-161558-ae4aa609-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js [hunk 12]
```diff
@@ -221,11 +246,17 @@
...
+            css("display", "flex") &&
+            css("display", "grid") &&
+            css("selector(test)") &&
+            css("selector(:is(*):where(*))");
+            return true;
+        } else {
```
Comment (cockpituous) [inline pkg/static/login.js:257]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221017-161558-ae4aa609-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
Pushed with a test which claims the browser has no CSS features at all, shows the bypass page and logins in.

File: pkg/static/login.js [hunk 13]
```diff
@@ -164,17 +164,42 @@
     }
 
     function requisites() {
-        function disableLogin(name) {
+        function showBypass(bypass) {
+            if (bypass) {
```
Comment (cockpituous) [inline pkg/static/login.js:168]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221024-100226-c3521259-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js (hunk 9 above)
Comment (cockpituous) [inline pkg/static/login.js:186]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221024-100226-c3521259-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js (hunk 13 above)
Comment (cockpituous) [inline pkg/static/login.js:168]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221024-133408-3792979a-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/static/login.js (hunk 9 above)
Comment (cockpituous) [inline pkg/static/login.js:186]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221024-133408-3792979a-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by jelly — DISMISSED** [review]

File: pkg/static/login.js (hunk 9 above)
Comment (cockpituous) [inline pkg/static/login.js:186]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17774-20221025-081404-67cb2980-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by jelly — APPROVED** [review]

**Comment by garrett** [comment]
Thanks, @jelly! :tada:
